src/approaches/llm_probe_approach.py

```diff
@@ -0,0 +1,187 @@
+"""Large language model (LLM) based policy guided bilevel planning approach.
+
+Example command line:
+    export OPENAI_API_KEY=<your API key>
+    python src/main.py --approach llm_probe --seed 0 \
+        --strips_learner oracle \
+        --env pddl_blocks_procedural_tasks \
+        --num_train_tasks 3 \
+        --num_test_tasks 1 \
+        --debug
+
+Easier setting:
+    python src/main.py --approach llm_probe --seed 0 \
+        --strips_learner oracle \
+        --env pddl_easy_delivery_procedural_tasks \
+        --pddl_easy_delivery_procedural_train_min_num_locs 2 \
+        --pddl_easy_delivery_procedural_train_max_num_locs 2 \
+        --pddl_easy_delivery_procedural_train_min_want_locs 1 \
+        --pddl_easy_delivery_procedural_train_max_want_locs 1 \
+        --pddl_easy_delivery_procedural_train_min_extra_newspapers 0 \
+        --pddl_easy_delivery_procedural_train_max_extra_newspapers 0 \
+        --pddl_easy_delivery_procedural_test_min_num_locs 2 \
+        --pddl_easy_delivery_procedural_test_max_num_locs 2 \
+        --pddl_easy_delivery_procedural_test_min_want_locs 1 \
+        --pddl_easy_delivery_procedural_test_max_want_locs 1 \
+        --pddl_easy_delivery_procedural_test_min_extra_newspapers 0 \
+        --pddl_easy_delivery_procedural_test_max_extra_newspapers 0 \
+        --num_train_tasks 5 \
+        --num_test_tasks 10 \
+        --debug
+"""
+from __future__ import annotations
+from typing import Dict, FrozenSet, Iterator, List, Optional, Set, Tuple
+from predicators.src.approaches import ApproachFailure
+from predicators.src import utils
+from predicators.src.approaches.open_loop_llm_approach import \
+    OpenLoopLLMApproach
+from predicators.src.settings import CFG
+from predicators.src.structs import Any, GroundAtom, ParameterizedOption, \
+    Sequence, State, _GroundNSRT
+
+
+class LLMProbeApproach(OpenLoopLLMApproach):
+    """LLMProbeApproach definition."""
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "llm_probe"
+    #
+    def _predict(self, state: State, atoms: Set[GroundAtom],
+                 goal: Set[GroundAtom], memory: Dict) -> _GroundNSRT:
+        # If we already have an abstract plan, execute the next step.
+        if "abstract_plan" in memory and memory["abstract_plan"]:
+            return memory["abstract_plan"].pop(0)
+        # Otherwise, we need to make a new abstract plan.
```

Comment:
this looks better, thanks! to finish off the refactoring, please change the parent class (`OpenLoopLLMApproach`) as well, so that the only thing that differs between the parent class and this class is the implementation of `_run_planner` (and now this class doesn't need to override `predict` anymore). the objective is to remove the lines of code that are in common with the parent class.

Reply:
and btw one of the reasons to remove duplicate code is that if we want to change it in the future, we only have to change it in one place, rather than two places